ext/xml/xml.c

```diff
@@ -1036,187 +1044,237 @@ PHP_FUNCTION(xml_parser_create_ns)
-	zval *pind, *mythis;
-
-	if (zend_parse_parameters(ZEND_NUM_ARGS(), "Oo", &pind, xml_parser_ce, &mythis) == FAILURE) {
-		RETURN_THROWS();
-	}
-
-	parser = Z_XMLPARSER_P(pind);
-
-	zval_ptr_dtor(&parser->object);
-	ZVAL_OBJ_COPY(&parser->object, Z_OBJ_P(mythis));
-
-	RETURN_TRUE;
-}
-/* }}} */
-
-/* {{{ Set up start and end element handlers */
-PHP_FUNCTION(xml_set_element_handler)
-{
-	xml_parser *parser;
-	zval *pind, *shdl, *ehdl;
-
-	if (zend_parse_parameters(ZEND_NUM_ARGS(), "Ozz", &pind, xml_parser_ce, &shdl, &ehdl) == FAILURE) {
-		RETURN_THROWS();
+static bool php_xml_check_string_method_arg(
+	unsigned int arg_num,
+	zend_object *object,
+	zend_string *method_name,
+	zend_fcall_info_cache *const parser_handler_fcc
+) {
+	if (ZSTR_LEN(method_name) == 0) {
+		ZEND_ASSERT(arg_num != 0);
+		/* Unset handler */
+		return true;
+	}
+
+	if (!object) {
+		ZEND_ASSERT(arg_num != 0);
+		zend_argument_value_error(arg_num, "an object must be set via xml_set_object() to be able to lookup method");
+		return false;
+	}
+
+	zend_class_entry *ce = object->ce;
+	zend_string *lc_name = zend_string_tolower(method_name);
+	zend_function *method_ptr = zend_hash_find_ptr(&ce->function_table, lc_name);
+	zend_string_release_ex(lc_name, 0);
+	if (!method_ptr) {
+		if (arg_num) {
+			zend_argument_value_error(arg_num, "method %s::%s() does not exist", ZSTR_VAL(ce->name), ZSTR_VAL(method_name));
+		}
+		return false;
 	}
 
-	parser = Z_XMLPARSER_P(pind);
-	xml_set_handler(&parser->startElementHandler, shdl);
-	xml_set_handler(&parser->endElementHandler, ehdl);
-	XML_SetElementHandler(parser->parser, _xml_startElementHandler, _xml_endElementHandler);
+	parser_handler_fcc->function_handler = method_ptr;
+	/* We set the calling scope to NULL to be able to differentiate a "method" set from a proper callable */
+	parser_handler_fcc->calling_scope = NULL;
+	parser_handler_fcc->called_scope = ce;
+	parser_handler_fcc->object = object;
 
-	RETURN_TRUE;
+	return true;
 }
-/* }}} */
 
-/* {{{ Set up character data handler */
-PHP_FUNCTION(xml_set_character_data_handler)
-{
-	xml_parser *parser;
-	zval *pind, *hdl;
-
-	if (zend_parse_parameters(ZEND_NUM_ARGS(), "Oz", &pind, xml_parser_ce, &hdl) == FAILURE) {
-		RETURN_THROWS();
+#define PHP_XML_CHECK_NEW_THIS_METHODS(parser_to_check, new_this_obj, fcc_field, handler_set_method) \
+	if ( \
+		ZEND_FCC_INITIALIZED(parser_to_check->fcc_field) \
+		&& parser_to_check->fcc_field.object == parser_to_check->object \
+		&& parser_to_check->fcc_field.calling_scope == NULL \
+	) { \
+		zend_string *method_name = zend_string_copy(parser_to_check->fcc_field.function_handler->common.function_name); \
+		zend_fcc_dtor(&parser_to_check->fcc_field); \
+		bool status = php_xml_check_string_method_arg(0, new_this_obj, method_name, &parser_to_check->fcc_field); \
+		if (status == false) { \
+			zend_argument_value_error(2, "cannot safely swap to object of class %s as method \"%s\" does not exist which was set via " handler_set_method, \
```

Comment:
```suggestion
			zend_argument_value_error(2, "cannot safely swap to object of class %s as method \"%s\" does not exist, which was set via " handler_set_method, \
```
makes it easier to read